File: examples/23-media-embed/main.go
package main

import (
	"encoding/base64"
	"fmt"
	"log"
	"os"
	"path/filepath"

	"github.com/djinn-soul/gopptx/pkg/pptx"
	"github.com/djinn-soul/gopptx/pkg/pptx/editor"
)

const (
	outputDir  = "examples/output"
	outputFile = "23_media_embed_editor.pptx"
)

func main() {
	if err := run(); err != nil {
		fmt.Fprintf(os.Stderr, "error: %v\n", err)
		os.Exit(1)
	}
}

func run() error {
	if err := os.MkdirAll(outputDir, 0o750); err != nil {
		return fmt.Errorf("create output dir: %w", err)
	}

	tmpDir, err := os.MkdirTemp("", "gopptx-media-embed-*")
	if err != nil {
		return fmt.Errorf("create temp dir: %w", err)
	}
	defer os.RemoveAll(tmpDir)

	basePath := filepath.Join(tmpDir, "media_embed_base.pptx")
	outPath := filepath.Join(outputDir, outputFile)

	baseSlides := []pptx.SlideContent{
		pptx.NewSlide("Media Embed (Editor)").
			AddBullet("Video, audio, and OLE media relationships inserted by PresentationEditor.").
			AddBullet("Provide real fixture files in examples/assets/23 to enable inserts."),
	}
	if err := pptx.WriteFile(basePath, "Media Embed Demo", baseSlides); err != nil {
		return fmt.Errorf("write base deck: %w", err)
	}

	ed, err := pptx.OpenPresentationEditor(basePath)
	if err != nil {
		return fmt.Errorf("open editor: %w", err)
	}
	defer ed.Close()

	posterPath := filepath.Join("examples", "assets", "23", "poster.png")
	if _, err := ensureMediaFixture(posterPath); err != nil {
		log.Printf("Poster fixture missing (%s): %v", posterPath, err)
	}

	videoFixturePath := filepath.Join("examples", "assets", "23", "sample.mp4")
	if fileExists(videoFixturePath) && fileExists(posterPath) {
		videoOpts := editor.NewAutoPlayVideoPlaybackOptions().
			WithLoop(true).
			WithMuted(true).
			WithVolume(0).
			WithAltText("Demo launch reel")
		if _, err := ed.AddVideoFromFileWithPlaybackOptions(
			0,
			videoFixturePath,
			posterPath,
			"video/mp4",
			videoOpts,
			600000,
			1400000,
			3600000,
			2100000,
		); err != nil {
			return fmt.Errorf("add video from fixture: %w", err)
		}
		log.Printf("Added video fixture: %s\n", videoFixturePath)
	} else {
		log.Printf("Video fixtures missing (need sample.mp4 + poster.png in examples/assets/23); skipping video insertion.")
	}

	audioMP3FixturePath := filepath.Join("examples", "assets", "23", "sample.mp3")
	audioWAVFixturePath := filepath.Join("examples", "assets", "23", "sample.wav")
	switch {
	case fileExists(audioMP3FixturePath):
		audioOpts := editor.NewAutoPlayAudioPlaybackOptions().
			WithLoop(true).
			WithPlayAcrossSlides(true).
			WithVolume(85).
			WithAltText("Background soundtrack")
		if _, err := ed.AddAudioFromFileWithPlaybackOptions(
			0,
			audioMP3FixturePath,
			"audio/mpeg",
			audioOpts,
			4500000,
			1500000,
			1800000,
			900000,
		); err != nil {
			return fmt.Errorf("add audio from mp3 fixture: %w", err)
		}
		log.Printf("Added audio fixture: %s\n", audioMP3FixturePath)
	case fileExists(audioWAVFixturePath):
		audioOpts := editor.NewAudioPlaybackOptions().
			WithVolume(85).
			WithAltText("Background soundtrack")
		if _, err := ed.AddAudioFromFileWithPlaybackOptions(
			0,
			audioWAVFixturePath,
			"audio/wav",
			audioOpts,
			4500000,
			1500000,
			1800000,
			900000,
		); err != nil {
			return fmt.Errorf("add audio from wav fixture: %w", err)
		}
		log.Printf("Added audio fixture: %s\n", audioWAVFixturePath)
	default:
		if _, err := ed.AddAudioWithIcon(0, tinyWAV(), tinyPNG(), "audio/wav", 4500000, 1500000, 1800000, 900000); err != nil {
			return fmt.Errorf("add fallback audio with icon: %w", err)
		}
		log.Printf(
			"Audio fixtures missing (sample.mp3/sample.wav); inserted fallback tiny WAV with generated icon.",
		)
	}

	oleFixturePath := filepath.Join("examples", "assets", "23", "sample_ole.bin")
	if fileExists(oleFixturePath) && fileExists(posterPath) {
		oleData, readErr := os.ReadFile(oleFixturePath)
		if readErr != nil {
			return fmt.Errorf("read ole fixture: %w", readErr)
		}
		if _, err := ed.AddOLEObject(0, oleData, tinyPNG(), "Excel.Sheet.12", 4500000, 2800000, 2100000, 1400000); err != nil {
			return fmt.Errorf("add ole object: %w", err)
		}
		log.Printf("Added OLE fixture: %s\n", oleFixturePath)
	} else {
		log.Printf("OLE fixtures missing (need sample_ole.bin + poster.png in examples/assets/23); skipping OLE insertion.")
	}

	if err := ed.Save(outPath); err != nil {
		return fmt.Errorf("save output: %w", err)
	}

	log.Printf("Generated media embed example: %s\n", outPath)
	return nil
}

func tinyPNG() []byte {
	const b64 = "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mP8Xw8AAoMBgA3FoXwAAAAASUVORK5CYII="
	data, err := base64.StdEncoding.DecodeString(b64)
	if err != nil {
		panic(err)
	}
	return data
}

func tinyWAV() []byte {
	return []byte{
		'R', 'I', 'F', 'F',
		0x25, 0x00, 0x00, 0x00,
		'W', 'A', 'V', 'E',
		'f', 'm', 't', ' ',
		0x10, 0x00, 0x00, 0x00,
		0x01, 0x00,
		0x01, 0x00,
		0x40, 0x1F, 0x00, 0x00,
		0x40, 0x1F, 0x00, 0x00,
		0x01, 0x00,
		0x08, 0x00,
		'd', 'a', 't', 'a',
		0x01, 0x00, 0x00, 0x00,
		0x80,
	}
}

func fileExists(filePath string) bool {
	_, err := os.Stat(filePath)
	return err == nil
}

func ensureMediaFixture(filePath string) (os.FileInfo, error) {
	return os.Stat(filePath)
}

File: examples/23-media-embed/main.py
"""Demonstrate embedding media (video, audio) into a PPTX presentation.

This example demonstrates:
- Adding a video file to a slide with add_video()
- Adding an audio file to a slide with add_audio()
- Falling back gracefully when media fixture files are not present
"""

from __future__ import annotations

from pathlib import Path

from gopptx import Presentation
from gopptx.schemas import Inches

_ASSETS_DIR = Path("examples/assets/23")
_VIDEO_PATH = _ASSETS_DIR / "sample.mp4"
_AUDIO_PATH = _ASSETS_DIR / "sample.mp3"


def _add_overview_slide(prs: Presentation) -> None:
    """Add an overview slide."""
    prs.add_bullet_slide(
        "Media Embed Demo",
        [
            "Video, audio, and OLE media can be embedded in PPTX.",
            "Provide real fixture files in examples/assets/23 to enable inserts.",
            "add_video(slide_idx, path, x, y, w, h)",
            "add_audio(slide_idx, path, x, y, w, h)",
        ],
    )


def _try_add_video(prs: Presentation) -> None:
    """Add a video slide if the fixture file exists."""
    prs.add_slide("Video Embed")
    idx = prs.slide_count - 1

    if _VIDEO_PATH.exists():
        prs.add_video(
            idx,
            str(_VIDEO_PATH),
            (Inches(1), Inches(1.5), Inches(8), Inches(4.5)),
        )
        print(f"  Added video from: {_VIDEO_PATH}")
    else:
        prs.add_textbox(
            idx,
            Inches(1),
            Inches(2),
            Inches(8),
            Inches(1),
            text=f"Video fixture not found: {_VIDEO_PATH}",
        )
        print(f"  Video fixture missing ({_VIDEO_PATH}); placeholder added.")


def _try_add_audio(prs: Presentation) -> None:
    """Add an audio slide if the fixture file exists."""
    prs.add_slide("Audio Embed")
    idx = prs.slide_count - 1

    if _AUDIO_PATH.exists():
        prs.add_audio(
            idx,
            str(_AUDIO_PATH),
            (Inches(4.5), Inches(1.5), Inches(1.8), Inches(0.9)),
        )
        print(f"  Added audio from: {_AUDIO_PATH}")
    else:
        prs.add_textbox(
            idx,
            Inches(1),
            Inches(2),
            Inches(8),
            Inches(1),
            text=f"Audio fixture not found: {_AUDIO_PATH}",
        )
        print(f"  Audio fixture missing ({_AUDIO_PATH}); placeholder added.")


def main() -> None:
    """Create a presentation demonstrating media embedding."""
    output_dir = Path("examples/output")
    output_dir.mkdir(exist_ok=True)

    with Presentation.new("Media Embed Demo") as prs:
        _add_overview_slide(prs)
        _try_add_video(prs)
        _try_add_audio(prs)

        output_path = output_dir / "23-media-embed.pptx"
        prs.save(str(output_path))
        print(f"Saved: {output_path}")

    print("\n=== SUMMARY ===")
    print("Demonstrated media embedding:")
    print("  - Video with add_video()")
    print("  - Audio with add_audio()")
    print("  - Graceful fallback when fixtures are missing")


if __name__ == "__main__":
    main()
